test/common/upstream/health_check_fuzz.cc

```diff
@@ -221,6 +260,215 @@ void TcpHealthCheckFuzz::raiseEvent(const Network::ConnectionEvent& event_type,
   }
 }
 
+void GrpcHealthCheckFuzz::allocGrpcHealthCheckerFromProto(
+    const envoy::config::core::v3::HealthCheck& config) {
+  health_checker_ = std::make_shared<TestGrpcHealthCheckerImpl>(
+      *cluster_, config, dispatcher_, runtime_, random_,
+      HealthCheckEventLoggerPtr(event_logger_storage_.release()));
+  ENVOY_LOG_MISC(trace, "Created Test Grpc Health Checker");
+}
+
+void GrpcHealthCheckFuzz::initialize(test::common::upstream::HealthCheckTestCase input) {
+  allocGrpcHealthCheckerFromProto(input.health_check_config());
+  cluster_->prioritySet().getMockHostSet(0)->hosts_ = {
+      makeTestHost(cluster_->info_, "tcp://127.0.0.1:80")};
+  expectSessionCreate();
+  expectStreamCreate(0);
+  health_checker_->start();
+  ON_CALL(runtime_.snapshot_, getInteger("health_check.min_interval", _))
+      .WillByDefault(testing::Return(45000));
+
+  if (DurationUtil::durationToMilliseconds(input.health_check_config().initial_jitter()) != 0) {
+    test_sessions_[0]->interval_timer_->invokeCallback();
+  }
+  if (input.health_check_config().has_reuse_connection()) {
+    reuse_connection_ = input.health_check_config().reuse_connection().value();
+  }
+}
+
+// Logic from respondResponseSpec() in unit tests
+void GrpcHealthCheckFuzz::respond(test::common::upstream::GrpcRespond grpc_respond) {
+  if (!test_sessions_[0]->timeout_timer_->enabled_) {
+    ENVOY_LOG_MISC(trace, "Timeout timer is disabled. Skipping response.");
+    return;
+  }
+  // These booleans help figure out when to end the stream
+  const bool has_data = grpc_respond.has_grpc_respond_bytes();
+  // TODO: Should I hardcode trailers? Particularly grpc-status = 0?
+  const bool has_trailers = grpc_respond.has_grpc_respond_trailers() &&
+                            grpc_respond.grpc_respond_trailers().has_trailers() &&
+                            grpc_respond.grpc_respond_trailers().trailers().headers_size() != 0;
+
+  ENVOY_LOG_MISC(trace, "Has data: {}. Has trailers: {}.", has_data, has_trailers);
+
+  bool end_stream_on_headers = !has_data && !has_trailers;
+
+  std::unique_ptr<Http::TestResponseHeaderMapImpl> response_headers =
+      std::make_unique<Http::TestResponseHeaderMapImpl>(
+          Fuzz::fromHeaders<Http::TestResponseHeaderMapImpl>(
+              grpc_respond.grpc_respond_headers().headers(), {}, {}));
+
+  response_headers->setStatus(
+      grpc_respond.grpc_respond_headers().status()); // 200 Required for gRPC
```

Comment:
probably want to remove this comment (or clarify)

Reply:
Removed